Fix: coerce unknown types to O dtype

brynpickering opened the pull request:
Added a check for Python objects that neither have numpy dtype `dtype` attributes nor are one of the builtin types so that type promotion succeeds and defaults to `object` dtype.

This deals with cases where user-defined instances are in the array which would otherwise trip up `numpy.result_type`.

I did think about placing this in `xarray.core.dtypes.preprocess_types` but chose to place it closest to where it causes issues (namely, on calling `xp.result_type`).

<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #9755 
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

brynpickering committed:
Coerce unknown types to O dtype

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

max-sixty commented:
do you think we could add a quick comment in the code describing what it's doing, for the less informed of us?

(does anyone know this better?)

kmuehlbauer committed:
Merge branch 'main' into fix-obj-dtype-infer

brynpickering committed:
Add comment on setting obj dtype

brynpickering commented:
Sure, done.

Another options for this would be to check all the args passed to the `compat.array_api_compat.result_type` method and if any of them are not a numpy-compatible object or a builtin type then just return `np.object_` directly. Passing the args to `np.result_type` is unnecessary as any object dtype in the list will automatically lead to an object dtype being returned (as it's the lowest common denominator).

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

brynpickering commented:
@keewis is there any timeline for this to be reviewed? We're currently stuck with xarray v2024.2.0 as a dependency until this is resolved which is starting to cause issues when used in some working environments.

keewis commented on xarray/compat/array_api_compat.py:
```diff
@@ -38,10 +38,18 @@ def _future_array_api_result_type(*arrays_and_dtypes, xp):
+        hasattr(t, "dtype") and isinstance(t.dtype, np.dtype) for t in arrays_and_dtypes
+    ]
+    if xp is np or any(is_np_dtype):
+        # Numpy can only apply type promotion rules on non-builtin & numpy-compatible Python objects.
+        # So we convert any incompatible objects (e.g. user-defined class instances) stored in the array to numpy Object dtype.
+        # Then, numpy's type promotion will fall back to Object dtype rather than raising an exception.
+        all_valid_arrays_and_dtypes = [
+            t if is_np_dtype[i] or t.__class__.__module__ == "builtins" else np.object_
+            for i, t in enumerate(arrays_and_dtypes)
+        ]
+        return xp.result_type(*all_valid_arrays_and_dtypes)
```
I wonder if we should just return `object` directly if there's an expected object dtype (since `result_type` should always return `object` in that case). Something like
```suggestion
        builtin_types = (bool, int, float, complex, str, bytes, dt.datetime, dt.timedelta)
        if any(not is_numpy and not isinstance(t, builtin_types) for is_numpy, t in zip(is_np_dtype, arrays_and_types)):
            return np.object_
        return xp.result_type(*arrays_and_dtypes)
```

keewis reviewed (commented):
sorry, this had completely dropped off my radar.

If this is about scalars, note that you can already wrap your custom object in a 0D array of object dtype to work around this (so you don't have to wait on us)

@shoyer, do you have an opinion on this?

shoyer reviewed (approved):
Let's get this in after applying @keewis's suggested clean-up?

brynpickering committed:
Update xarray/compat/array_api_compat.py

Co-authored-by: Justus Magin <keewis@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

keewis committed:
import datetime

keewis committed:
typo

keewis committed:
use `zip` with `strict=True`

keewis committed:
refactor the dtype checks into functions

keewis committed:
return a dtype object instead of the dtype type

jsignell committed:
Merge branch 'main' into fix-obj-dtype-infer

jsignell commented:
@keewis is looks like this one is good to merge. Right? Just needs a note in "what's new"

brynpickering commented:
@jsignell this may have been superseded by #10423? It certainly seems to be at odds with #10423 since a bunch of the extension array tests are now failing. Let me see if removing the implementation here introduced here but keeping the tests (and checking against the MWE I raised in the issue) works as expected.

keewis committed:
revert the changes to `result_type`

brynpickering committed:
Update given TypeError fallback

brynpickering commented:
@keewis I've removed the two methods previously added in this PR, in favour of the TypeError fallback. I've then removed `try/except` statements where they would no longer reach a TypeError since the method being called handles that itself. I don't think it makes sense to try merging the two `object` type promotion methods since one is there to override default numpy type promotion and the other is a fallback for types that numpy would otherwise raise an error for on attempting to promote.

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

brynpickering commented:
@keewis @shoyer is there any way I can help push this forward? Would love to be able to update our project to latest xarray!

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

keewis committed:
Merge branch 'main' into fix-obj-dtype-infer

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

shoyer commented on xarray/compat/array_api_compat.py:
```diff
@@ -1,7 +1,20 @@
+    bytes,
+    dt.datetime,
+    dt.timedelta,
+)
```
It does not look like this is used in the current version of this PR

brynpickering replied:
Good point. Removed.

shoyer commented on xarray/compat/array_api_compat.py:
```diff
@@ -38,12 +51,15 @@ def _future_array_api_result_type(*arrays_and_dtypes, xp):
+        else:
+            return _future_array_api_result_type(*arrays_and_dtypes, xp=xp)
+    except TypeError:
+        return np.dtype(object)
```
This entire module is intended as forwards compatible replacements for array API functionality, so we shouldn't add behavior here that deviates from `result_type` in the array API. I would much rather prefer that this changed behavior be done in Xarray's specific `dtypes.result_type()`. (Also, it should have a brief comment explaining why `TypeError` mean object dtype is appropriate.)

brynpickering replied:
Fair enough. I've moved it up to `dtypes.result_type()`

shoyer commented on xarray/core/utils.py:
```diff
@@ -213,7 +213,7 @@ def maybe_coerce_to_str(index, original_coords):
 
     try:
         result_type = dtypes.result_type(*original_coords)
-    except (TypeError, ValueError):
+    except ValueError:
         pass
```
Does this change also need test coverage? I don't see how `results.result_type()` would cover it.

brynpickering replied:
I've added a test

shoyer reviewed (commented):
Sorry for neglecting this one for so long!

brynpickering committed:
Fixes post-review

- Add `maybe_coerce_to_str` test
- Remove redundant builtins list
- Add typeerror comment

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brynpickering committed:
Merge branch 'main' into fix-obj-dtype-infer

shoyer commented on xarray/core/dtypes.py:
```diff
@@ -326,8 +320,16 @@ def result_type(
+    except TypeError:
+        # unknown python objects will raise a TypeError in `xp.result_type`;
+        # We assume the user wants them to be there and therefore promote to object dtype instead of raising.
+        promote_to_object = True
```
TypeError is a really common exception, so catching any TypeError from should_promote_to_object this could hide other errors. I would prefer to scope this as tightly as possible around `result_type()`.

Could you create a custom error type to raise there? Something like:
```python
        arr_or_dtype = maybe_promote_to_variable_width(arr_or_dtype)
        try:
            result_type = array_api_compat.result_type(arr_or_dtype, xp=xp)
        except TypeError:
            raise ResultTypeError
```

and then you can catch ResultTypeError explicitly here, instead of any TypeError

brynpickering committed:
Fix when typerror is caught

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shoyer commented:
@brynpickering Does the custom exception subclass make sense to you? I just want to be really careful about catching broad exceptions like TypeError because otherwise this can cover up bugs.

brynpickering commented:
@shoyer it makes sense to be careful about the error, although I do wonder what cases there are for TypeErrors in this particular instance which _aren't_ ones where you'd want to just return an `object` dtype. I know that not being able to think of cases isn't the same as there being no cases, of course!
